lib/thread/freertos/semaphore.cpp

```diff
@@ -5,45 +5,41 @@
 #include <platform/irq_manager.hpp>
 
 ecl::semaphore::semaphore()
-    :m_semaphore{nullptr}
+        :m_semaphore{}
+        ,m_cnt{0}
 {
-    // TODO: find better way to implement semaphore, rather than via
-    // xSemaphoreCreateCounting(). Problem is that xSemaphoreCreateCounting()
-    // limits maximum count of a semaphore. See FreeRTOS docs for details.
-    m_semaphore = xSemaphoreCreateCounting(10, 0);
-    ecl_assert(m_semaphore);
 }
 
 ecl::semaphore::~semaphore()
 {
-    vSemaphoreDelete(m_semaphore);
 }
 
 void ecl::semaphore::signal()
 {
-    // HACK:
-    // Give operation fails if there is no more room in FreeRTOS queue.
-    // Idea is to postpone a giving until consumer will free some space for us.
-    if (!ecl::irq_manager::in_isr()) {
-        while (xSemaphoreGive(m_semaphore) != pdTRUE) {
-            this_thread::sleep_for(100);
-        }
-    } else {
-        // Can't block an ISR, event will be lost
-        xSemaphoreGiveFromISR(m_semaphore, nullptr);
+    if (m_cnt.fetch_add(1) >= -1) {
+        m_semaphore.signal();
     }
 }
 
 void ecl::semaphore::wait()
 {
-    auto rc = xSemaphoreTake(m_semaphore, portMAX_DELAY);
-    ecl_assert(rc == pdTRUE);
+    if (m_cnt.fetch_sub(1) <= 0) {
+        m_semaphore.wait();
+    }
 }
 
 ecl::err ecl::semaphore::try_wait()
 {
-    auto rc = xSemaphoreTake(m_semaphore, 0);
-    return rc == pdTRUE ? ecl::err::ok : ecl::err::again;
+    int saved;
+    if ((saved = m_cnt.fetch_sub(1)) >= 0) {
+        return err::ok; // Got semaphore.
+    } else if (saved == -1) {
+        // Unblock any poor thread that was blocked just right now.
+        m_semaphore.signal();
```

Comment:
m_cnt is 0.

Thread 1:
- try_wait()
- fetch_sub return -1

Preemption

Thread 2:
- wait()
- fetch_sub returns -2
- wait on m_semaphore

Preemption

Thread 1:
- fetch_sub returned -1, so go signal m_semaphore
- thread 2 is unblocked and holds the semaphore

That's a bug. Semaphore was never signaled.